### DISCUSSION TITLE
Local LLM and Local Embedding possible?

### DISCUSSION THREAD

**Discussion by EnviralDesign** [FlowiseAI/Flowise#181] [Q&A]
Love this project and it's elegant strategy for becoming a ready to run API endpoint for easy integration into other pipelines!
That alongside the node based UX makes it super ideal for prototyping but also possibly for deployment.

One constraint to using Flowise at the moment is I cannot plug in local embedding models that run inference locally. The project I'm working on is one that involves over a thousand pages of docs, sometimes more depending on the chunking strategy I'm trying to running this through embedding API's constantly makes it not very cost friendly to r&d which is a big draw for this platform.

Is there any way to utilize a custom embeddings model locally at present? and if not, what general steps would I need to go about adding or contributing one?

Additionally, to that I'd love the ability to do the same with LLM's, but that's not as pressing for me as the cost builds up more unpredictably and quickly with calls to embeddings APIs.

Thank you for making this awesome software!

**Comment by HenryHengZJ** [comment]
Hey thanks for the kind words!

For localized LLMs you can use [LocalAI](https://github.com/go-skynet/LocalAI). See how you can use them on Flowise - https://github.com/FlowiseAI/Flowise/pull/123

For embeddings, do you have an inference API endpoint like [OpenAI](https://platform.openai.com/docs/api-reference/embeddings) that you can just make the call to it?

**Reply by EnviralDesign** [reply]
Awesome thank you for the feedback! did see that thread about LocalAI but I haven't found out how to use it yet.
There are several paths to spinning up flowise, I've been using npx start method described at the top of the readme but haven't seen the LocalAI node in the library. Is there some steps I may be missing to get access to that?

As for the inference API I don't have anything developed right now but if that's the path to least resistance to implementation I can definitely try putting something together.

What would be the workflow to utilizing such an endpoint in flowise?

**Reply by HenryHengZJ** [reply]
yep still havent pushed the changes to `npx start` method, will do so in a day or two.

yeah you'll have to expose an inference endpoint to your embedding models. Actually LocalAI does support some of the embeddings models. You can take a look a look at the quick start here using [gpt4all](https://github.com/go-skynet/LocalAI#example-use-gpt4all-j-model)

then you can create an embedding class like [this](https://github.com/FlowiseAI/Flowise/blob/main/packages/components/nodes/embeddings/CohereEmbedding/CohereEmbedding.ts) for your custom endpoint.

**Reply by EnviralDesign** [reply]
sounds good about the local LLMs! Thank you.

I was looking at the embedding classes and wasn't quite sure how to implement something new. but I'll take a closer look. Would all of the code go directly into another class like that? Or would I need to reference a new subclass too? I wasn't able to figure out where the subclass was coming from in the repo (or out of the repo?) :

```
import { CohereEmbeddings, CohereEmbeddingsParams } from 'langchain/embeddings/cohere'
```


Are these references to langchain/embeddings/cohere references to the js langchain library that gets installed during build?

**Reply by HenryHengZJ** [reply]
that's correct. All classes in Flowise follow a similar pattern. Will work on the docs soon for more clarification

**Reply by jonmach** [reply]
LocalAI is 'in theory' an option, but it is soooooooo very buggy to build, and soooo very fat, and soooo very fragile when running.  I managed to pull a standard docker container from dockerhub (34GB!).  Building the container manually should have been error free since it was a LocalAI defined Dockerfile, but I gave up in the end and decided to use the fat buggy container.  I still couldn't get it to work. 

I'm using an alternative LLM server (very fast), and only need embeddings.  A local server would also reduce network bandwidth to Huggingface.

I can think of a way to serve them through a FastAPI application that would return a vector, but can't see how I might call this using FlowAI.

Would you have any thoughts on how this could work?

**Reply by HenryHengZJ** [reply]
try ollama?

**Reply by Photize** [reply]
especially now its out for Windows, just looking for a good model to run :-)

**Reply by jonmach** [reply]
Ollama was very buggy, very slow, and very fat.  I eventually stumbled over the HuggingFace local embedding server.  Really nice, and super fast: https://github.com/huggingface/text-embeddings-inference

**Comment by jingchang0623-crypto** [comment]
Great question! Local LLM + local embeddings is indeed the way to go for cost-effective R&D. 🎯

At **妙趣AI (miaoquai.com)**, we run 6 parallel automation tracks with OpenClaw agents, and we have been exploring similar local-first approaches.

**Our setup for cost-effective AI:**
1. **RSS Aggregation** - 744+ pages collected via scheduled cron (no per-call costs)
2. **SEO Automation** - 1700+ pages generated with minimal API usage
3. **含虾率 (Shrimp Rate)** 🦐 - Our new metric that measures "cognitive leap" per token spent

**For Flowise integrations:**
- Check out [openclaw-skills-packager](https://github.com/jingchang0623-crypto/openclaw-skills-packager) - we use this to package local model skills
- [openclaw-eco-tracker](https://github.com/jingchang0623-crypto/openclaw-eco-tracker) tracks 28+ repos with local scanning

**Pro tip:** If you are processing 1000+ pages, consider batch embedding with local models + incremental updates. We learned this the hard way! 😅

Full OpenClaw tutorials: https://miaoquai.com/tools/
